File: shared/types/src/messages.ts
import {
  ErrorCode,
  GripperCommand,
  PeerMode,
  PROTOCOL_VERSION,
  Role,
  RobotMode,
  SessionState,
} from "./enums";

/** Fields present on every message (mirrors teleop_protocol.messages.BaseMessage). */
export interface BaseMessage {
  protocol_version: number;
  type: string;
  session_id: string;
  timestamp: number;
}

export const PROTOCOL_VERSION_DEFAULT = PROTOCOL_VERSION;

// ---------------------------------------------------------------------------
// Session messages
// ---------------------------------------------------------------------------

export interface SessionJoin extends BaseMessage {
  type: "session.join";
  role: Role;
  operator_id?: string | null;
  token?: string | null;
  display_name?: string | null;
}

export interface SessionJoined extends BaseMessage {
  type: "session.joined";
  role: Role;
  participant_id: string;
  operator_id?: string | null;
  session_state: SessionState;
  owner_operator_id?: string | null;
  participants: Record<string, number>;
}

export interface SessionLeave extends BaseMessage {
  type: "session.leave";
  participant_id?: string | null;
  reason?: string | null;
}

export interface SessionError extends BaseMessage {
  type: "session.error";
  code: ErrorCode;
  message: string;
}

// ---------------------------------------------------------------------------
// Ownership messages
// ---------------------------------------------------------------------------

export interface TeleopClaim extends BaseMessage {
  type: "teleop.claim";
  operator_id: string;
}

export interface TeleopClaimed extends BaseMessage {
  type: "teleop.claimed";
  operator_id: string;
  granted: boolean;
  reason?: string | null;
}

export interface TeleopRelease extends BaseMessage {
  type: "teleop.release";
  operator_id: string;
}

export interface TeleopOwnership extends BaseMessage {
  type: "teleop.ownership";
  owner_operator_id?: string | null;
  teleop_state: SessionState;
}

// ---------------------------------------------------------------------------
// Teleop control messages
// ---------------------------------------------------------------------------

export interface TeleopEnable extends BaseMessage {
  type: "teleop.enable";
  operator_id: string;
  seq: number;
}

export interface TeleopDisable extends BaseMessage {
  type: "teleop.disable";
  operator_id: string;
  seq: number;
}

export interface TeleopClutch extends BaseMessage {
  type: "teleop.clutch";
  operator_id: string;
  seq: number;
  engaged: boolean;
}

export interface TeleopEstop extends BaseMessage {
  type: "teleop.estop";
  operator_id?: string | null;
  seq: number;
  engaged: boolean;
  reset: boolean;
}

export interface TeleopHome extends BaseMessage {
  type: "teleop.home";
  operator_id: string;
  seq: number;
}

export interface TeleopGripper extends BaseMessage {
  type: "teleop.gripper";
  operator_id: string;
  seq: number;
  command: GripperCommand;
}

// ---------------------------------------------------------------------------
// State messages
// ---------------------------------------------------------------------------

export interface TeleopLeaderState extends BaseMessage {
  type: "teleop.leader_state";
  operator_id: string;
  seq: number;
  leader_jointstates: number[];
  gripper: number;
  enabled: boolean;
  clutched: boolean;
}

export interface TeleopFollowerState extends BaseMessage {
  type: "teleop.follower_state";
  seq_ack: number;
  follower_jointstates: number[];
  robot_mode: RobotMode;
  estopped: boolean;
  clutched: boolean;
  packet_age_ms: number;
}

export interface TeleopStatus extends BaseMessage {
  type: "teleop.status";
  teleop_state: SessionState;
  owner_operator_id?: string | null;
  peer_mode: PeerMode;
  rtt_ms?: number | null;
  video_ok: boolean;
  data_ok: boolean;
}

export interface TeleopAck extends BaseMessage {
  type: "teleop.ack";
  ack_type: string;
  seq?: number | null;
  accepted: boolean;
  reason?: string | null;
}

export interface TeleopErrorMsg extends BaseMessage {
  type: "teleop.error";
  code: ErrorCode;
  message: string;
  related_seq?: number | null;
}

export interface TeleopHeartbeat extends BaseMessage {
  type: "teleop.heartbeat";
  sender: Role;
  seq: number;
}

export interface TeleopLatency extends BaseMessage {
  type: "teleop.latency";
  sender: Role;
  rtt_ms: number;
  measured_at: number;
}

// ---------------------------------------------------------------------------
// Signaling messages (WebRTC negotiation + session-state queries)
// ---------------------------------------------------------------------------

export interface WebRTCOffer extends BaseMessage {
  type: "webrtc.offer";
  to_role: Role;
  to_participant_id?: string | null;
  from_participant_id?: string | null;
  from_role?: Role | null;
  sdp: string;
}

export interface WebRTCAnswer extends BaseMessage {
  type: "webrtc.answer";
  to_role: Role;
  to_participant_id?: string | null;
  from_participant_id?: string | null;
  from_role?: Role | null;
  sdp: string;
}

export interface WebRTCIceCandidate extends BaseMessage {
  type: "webrtc.ice_candidate";
  to_role: Role;
  to_participant_id?: string | null;
  from_participant_id?: string | null;
  from_role?: Role | null;
  candidate?: string | null;
  sdp_mid?: string | null;
  sdp_mline_index?: number | null;
}

export interface SessionStateRequest extends BaseMessage {
  type: "session.state_request";
}

export interface SessionStateMessage extends BaseMessage {
  type: "session.state";
  teleop_state: SessionState;
  owner_operator_id?: string | null;
  estopped: boolean;
  participants: Record<string, number>;
}

/** Union of every message type in the protocol. */
export type TeleopMessage =
  | SessionJoin
  | SessionJoined
  | SessionLeave
  | SessionError
  | TeleopClaim
  | TeleopClaimed
  | TeleopRelease
  | TeleopOwnership
  | TeleopEnable
  | TeleopDisable
  | TeleopClutch
  | TeleopEstop
  | TeleopHome
  | TeleopGripper
  | TeleopLeaderState
  | TeleopFollowerState
  | TeleopStatus
  | TeleopAck
  | TeleopErrorMsg
  | TeleopHeartbeat
  | TeleopLatency
  | WebRTCOffer
  | WebRTCAnswer
  | WebRTCIceCandidate
  | SessionStateRequest
  | SessionStateMessage;

File: shared/protocol/teleop_protocol/messages.py
"""Pydantic models for every message exchanged over the signaling WebSocket
and the WebRTC data channels.

Every message extends :class:`BaseMessage`, which carries the four fields
required by the spec on every message: ``protocol_version``, ``type``,
``session_id`` and ``timestamp``.
"""

import time
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field

from .enums import ErrorCode, GripperCommand, PeerMode, Role, RobotMode, SessionState
from .version import PROTOCOL_VERSION


class BaseMessage(BaseModel):
    protocol_version: int = PROTOCOL_VERSION
    type: str
    session_id: str
    timestamp: float = Field(default_factory=lambda: time.time())


# ---------------------------------------------------------------------------
# Session messages
# ---------------------------------------------------------------------------


class SessionJoin(BaseMessage):
    type: Literal["session.join"] = "session.join"
    role: Role
    operator_id: Optional[str] = None
    token: Optional[str] = None
    display_name: Optional[str] = None


class SessionJoined(BaseMessage):
    type: Literal["session.joined"] = "session.joined"
    role: Role
    participant_id: str
    operator_id: Optional[str] = None
    session_state: SessionState
    owner_operator_id: Optional[str] = None
    participants: Dict[str, int] = Field(default_factory=dict)


class SessionLeave(BaseMessage):
    type: Literal["session.leave"] = "session.leave"
    participant_id: Optional[str] = None
    reason: Optional[str] = None


class SessionError(BaseMessage):
    type: Literal["session.error"] = "session.error"
    code: ErrorCode
    message: str


# ---------------------------------------------------------------------------
# Ownership messages
# ---------------------------------------------------------------------------


class TeleopClaim(BaseMessage):
    type: Literal["teleop.claim"] = "teleop.claim"
    operator_id: str


class TeleopClaimed(BaseMessage):
    type: Literal["teleop.claimed"] = "teleop.claimed"
    operator_id: str
    granted: bool
    reason: Optional[str] = None


class TeleopRelease(BaseMessage):
    type: Literal["teleop.release"] = "teleop.release"
    operator_id: str


class TeleopOwnership(BaseMessage):
    type: Literal["teleop.ownership"] = "teleop.ownership"
    owner_operator_id: Optional[str] = None
    teleop_state: SessionState


# ---------------------------------------------------------------------------
# Teleop control messages (operator -> follower, reliable channel)
# ---------------------------------------------------------------------------


class TeleopEnable(BaseMessage):
    type: Literal["teleop.enable"] = "teleop.enable"
    operator_id: str
    seq: int


class TeleopDisable(BaseMessage):
    type: Literal["teleop.disable"] = "teleop.disable"
    operator_id: str
    seq: int


class TeleopClutch(BaseMessage):
    type: Literal["teleop.clutch"] = "teleop.clutch"
    operator_id: str
    seq: int
    engaged: bool


class TeleopEstop(BaseMessage):
    type: Literal["teleop.estop"] = "teleop.estop"
    operator_id: Optional[str] = None
    seq: int
    engaged: bool = True
    reset: bool = False


class TeleopHome(BaseMessage):
    type: Literal["teleop.home"] = "teleop.home"
    operator_id: str
    seq: int


class TeleopGripper(BaseMessage):
    type: Literal["teleop.gripper"] = "teleop.gripper"
    operator_id: str
    seq: int
    command: GripperCommand


# ---------------------------------------------------------------------------
# State messages
# ---------------------------------------------------------------------------


class TeleopLeaderState(BaseMessage):
    type: Literal["teleop.leader_state"] = "teleop.leader_state"
    operator_id: str
    seq: int
    leader_jointstates: List[float]
    gripper: float
    enabled: bool
    clutched: bool


class TeleopFollowerState(BaseMessage):
    type: Literal["teleop.follower_state"] = "teleop.follower_state"
    seq_ack: int
    follower_jointstates: List[float]
    robot_mode: RobotMode
    estopped: bool
    clutched: bool
    packet_age_ms: float


class TeleopStatus(BaseMessage):
    type: Literal["teleop.status"] = "teleop.status"
    teleop_state: SessionState
    owner_operator_id: Optional[str] = None
    peer_mode: PeerMode = PeerMode.UNKNOWN
    rtt_ms: Optional[float] = None
    video_ok: bool = False
    data_ok: bool = False


class TeleopAck(BaseMessage):
    type: Literal["teleop.ack"] = "teleop.ack"
    ack_type: str
    seq: Optional[int] = None
    accepted: bool
    reason: Optional[str] = None


class TeleopError(BaseMessage):
    type: Literal["teleop.error"] = "teleop.error"
    code: ErrorCode
    message: str
    related_seq: Optional[int] = None


class TeleopHeartbeat(BaseMessage):
    type: Literal["teleop.heartbeat"] = "teleop.heartbeat"
    sender: Role
    seq: int


class TeleopLatency(BaseMessage):
    type: Literal["teleop.latency"] = "teleop.latency"
    sender: Role
    rtt_ms: float
    measured_at: float


# ---------------------------------------------------------------------------
# Signaling messages (WebRTC negotiation + session-state queries)
#
# These never cross a WebRTC data channel themselves -- they're exchanged
# over the signaling WebSocket between participants and the signaling
# server -- but gateways parse them with the same `parse_message` dispatcher
# as the data-channel messages above, so they're defined alongside them.
# ---------------------------------------------------------------------------


class WebRTCOffer(BaseMessage):
    type: Literal["webrtc.offer"] = "webrtc.offer"
    to_role: Role
    to_participant_id: Optional[str] = None
    from_participant_id: Optional[str] = None
    from_role: Optional[Role] = None
    sdp: str


class WebRTCAnswer(BaseMessage):
    type: Literal["webrtc.answer"] = "webrtc.answer"
    to_role: Role
    to_participant_id: Optional[str] = None
    from_participant_id: Optional[str] = None
    from_role: Optional[Role] = None
    sdp: str


class WebRTCIceCandidate(BaseMessage):
    type: Literal["webrtc.ice_candidate"] = "webrtc.ice_candidate"
    to_role: Role
    to_participant_id: Optional[str] = None
    from_participant_id: Optional[str] = None
    from_role: Optional[Role] = None
    candidate: Optional[str] = None
    sdp_mid: Optional[str] = None
    sdp_mline_index: Optional[int] = None


class SessionStateRequest(BaseMessage):
    type: Literal["session.state_request"] = "session.state_request"


class SessionStateMessage(BaseMessage):
    type: Literal["session.state"] = "session.state"
    teleop_state: SessionState
    owner_operator_id: Optional[str] = None
    estopped: bool = False
    participants: Dict[str, int] = Field(default_factory=dict)
